forge/src/executor/mod.rs

```diff
@@ -0,0 +1,353 @@
+    //inspector: LogCollector,
+}
+
+impl<DB> Executor<DB>
+where
+    DB: DatabaseRef,
+{
+    pub fn new(inner_db: DB, env: Env) -> Self {
+        Executor { db: CacheDB::new(inner_db), env }
+    }
+
+    /// Set the balance of an account.
+    pub fn set_balance(&mut self, address: Address, amount: U256) {
+        let mut account = self.db.basic(address);
+        account.balance = amount;
+
+        self.db.insert_cache(address, account);
+    }
+
+    /// Calls the `setUp()` function on a contract.
+    pub fn setup(
+        &mut self,
+        address: Address,
+    ) -> std::result::Result<(Return, Vec<RawLog>), EvmError> {
+        let CallResult { status, logs, .. } = self.call_committing::<(), _, _>(
+            Address::zero(),
+            address,
+            "setUp()",
+            (),
+            0.into(),
+            None,
+        )?;
+        Ok((status, logs))
+    }
+
+    /// Performs a call to an account on the current state of the VM.
+    ///
+    /// The state after the call is persisted.
+    pub fn call_committing<D: Detokenize, T: Tokenize, F: IntoFunction>(
+        &mut self,
+        from: Address,
+        to: Address,
+        func: F,
+        args: T,
+        value: U256,
+        abi: Option<&Abi>,
+    ) -> std::result::Result<CallResult<D>, EvmError> {
+        let func = func.into();
+        let calldata = Bytes::from(encode_function_data(&func, args)?.to_vec());
+        let RawCallResult { result, status, gas, logs, .. } =
+            self.call_raw_committing(from, to, calldata, value)?;
+        match status {
+            return_ok!() => {
+                let result = decode_function_data(&func, result, false)?;
+                Ok(CallResult { status, result, gas, logs, state_changeset: None })
+            }
+            _ => {
+                let reason = foundry_utils::decode_revert(result.as_ref(), abi)
+                    .unwrap_or_else(|_| format!("{:?}", status));
+                Err(EvmError::Execution {
+                    status,
+                    reason,
+                    gas_used: gas,
+                    logs,
+                    state_changeset: None,
+                })
+            }
+        }
+    }
+
+    /// Performs a raw call to an account on the current state of the VM.
+    ///
+    /// The state after the call is persisted.
+    pub fn call_raw_committing(
+        &mut self,
+        from: Address,
+        to: Address,
+        calldata: Bytes,
+        value: U256,
+    ) -> Result<RawCallResult> {
+        let mut evm = EVM::new();
+        evm.env = self.env.clone();
+        evm.env.tx = TxEnv {
+            caller: from,
+            transact_to: TransactTo::Call(to),
+            data: calldata,
+            value,
+            ..Default::default()
+        };
+        evm.database(&mut self.db);
+
+        // Run the call
+        let state = Rc::new(RefCell::new(ExecutorState::new()));
```

Comment:
I opted to use an `Rc<RefCell<ExecutorState>>` for the executor state since we will pass it to a "meta-inspector" later that will manage a list of other inspectors that all share the same state. Does this make sense?

Primary reason I didn't opt for an `Arc<Mutex<ExecutorState>>` is that all the VMs are single-threaded anyway